File: src/keld/diffview.py
from __future__ import annotations

import difflib
from pathlib import Path

from .console import console


def diff_lines(before: str | None, after: str, label: str) -> list[str]:
    return list(difflib.unified_diff(
        (before or "").splitlines(keepends=True),
        after.splitlines(keepends=True),
        fromfile=f"a/{label}", tofile=f"b/{label}",
    ))


def render(before: str | None, after: str, label: str | Path) -> None:
    for raw in diff_lines(before, after, str(label)):
        line = raw.rstrip("\n")
        if line.startswith("+") and not line.startswith("+++"):
            style = "green"
        elif line.startswith("-") and not line.startswith("---"):
            style = "red"
        elif line.startswith("@@"):
            style = "cyan"
        else:
            style = "dim"
        # markup=False: config lines contain brackets ([otel], JSON) that rich
        # would otherwise try to parse as markup.
        console.print(line, style=style, markup=False)

File: internal/diffview/diffview.go
package diffview

import (
	"fmt"
	"strings"

	"github.com/fatih/color"
	"github.com/pmezard/go-difflib/difflib"

	"github.com/ncx-ai/keld-cli/internal/console"
)

var (
	green = color.New(color.FgGreen)
	red   = color.New(color.FgRed)
	cyan  = color.New(color.FgCyan)
	dim   = color.New(color.Faint)
)

// Render prints a colorized unified diff of before→after to console.Out.
// before==nil is treated as empty (no prior content). label is used as the
// filename in the diff header (FromFile "a/label", ToFile "b/label").
func Render(before *string, after, label string) {
	beforeStr := ""
	if before != nil {
		beforeStr = *before
	}

	diffStr, err := difflib.GetUnifiedDiffString(difflib.UnifiedDiff{
		A:        difflib.SplitLines(beforeStr),
		B:        difflib.SplitLines(after),
		FromFile: "a/" + label,
		ToFile:   "b/" + label,
		Context:  3,
	})
	if err != nil {
		fmt.Fprintln(console.Out, err)
		return
	}

	for _, raw := range strings.Split(diffStr, "\n") {
		// Reconstruct the line without trailing newline (diffStr uses \n as
		// separator; Split already strips them).
		line := raw
		switch {
		case strings.HasPrefix(line, "+") && !strings.HasPrefix(line, "+++"):
			green.Fprintln(console.Out, line)
		case strings.HasPrefix(line, "-") && !strings.HasPrefix(line, "---"):
			red.Fprintln(console.Out, line)
		case strings.HasPrefix(line, "@@"):
			cyan.Fprintln(console.Out, line)
		default:
			dim.Fprintln(console.Out, line)
		}
	}
}
